File: packages/cli/templates/scripts/registry/file-utils.js
'use strict';

/**
 * file-utils.js
 *
 * Single Responsibility: file collection and path helpers shared across scanners.
 * No scanning logic, no schema building — only filesystem utilities.
 *
 * Skip-list sources merged inside collectFiles:
 *   - DEFAULT_IGNORE (universal: node_modules, .git, dist, ...)
 *   - explicit `ignore` argument
 *   - env `MUSTARD_SCAN_IGNORE` (comma-separated names)
 *   - directory entries parsed from the subproject's .gitignore (parseGitignoreDirs)
 *
 * Usage:
 *   const { collectFiles, relativePath, readFileSafe } = require('./registry/file-utils');
 */

const fs = require('fs');
const path = require('path');

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const DEFAULT_IGNORE = [
  'node_modules', 'bin', 'obj', 'dist', '.next',
  '__pycache__', '.venv', 'venv', 'target', 'build',
  '.git', 'migrations', 'Migrations',
];

// ---------------------------------------------------------------------------
// parseGitignoreDirs
// ---------------------------------------------------------------------------

/**
 * Extract directory-name patterns from a .gitignore string.
 *
 * Conservative parsing — keeps only entries that look like a plain folder name:
 *   - non-empty
 *   - no whitespace
 *   - no glob chars (*, ?, [, ])
 *   - no slashes (path-anchored entries are skipped)
 *   - not a negation (! prefix)
 *   - not a comment (# prefix)
 *
 * Trailing slashes are stripped. Leading slashes cause the entry to be skipped
 * (path-anchored — out of scope for the simple folder-name skip-list).
 *
 * @param {string} content - raw .gitignore file content
 * @returns {string[]} - list of folder names
 */
function parseGitignoreDirs(content) {
  if (typeof content !== 'string' || !content) return [];
  const out = [];
  for (const rawLine of content.split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line || line.startsWith('#') || line.startsWith('!')) continue;
    if (line.startsWith('/')) continue; // path-anchored
    if (/[\s*?\[\]]/.test(line)) continue; // glob or whitespace
    const name = line.replace(/\/$/, ''); // strip trailing slash before path check
    if (name.includes('/')) continue; // nested path, not a bare name
    out.push(name);
  }
  return out;
}

// ---------------------------------------------------------------------------
// collectFiles
// ---------------------------------------------------------------------------

/**
 * Recursively collect all files with the given extension under a directory.
 * Skips DEFAULT_IGNORE directories, dot-directories, and any extra ignore dirs.
 *
 * @param {string} dir - root directory to walk
 * @param {string} extension - file extension including dot, e.g. '.cs', '.ts'
 * @param {string[]} [ignore] - additional directory names to skip
 * @returns {string[]} - absolute file paths
 */
function collectFiles(dir, extension, ignore = []) {
  const envIgnore = String(process.env.MUSTARD_SCAN_IGNORE || '')
    .split(',')
    .map(s => s.trim())
    .filter(Boolean);

  let gitignoreDirs = [];
  try {
    const gi = path.join(dir, '.gitignore');
    if (fs.existsSync(gi)) {
      gitignoreDirs = parseGitignoreDirs(fs.readFileSync(gi, 'utf-8'));
    }
  } catch { /* fail-open */ }

  const allIgnore = new Set([...DEFAULT_IGNORE, ...ignore, ...envIgnore, ...gitignoreDirs]);
  const results = [];

  function walk(currentDir) {
    try {
      const entries = fs.readdirSync(currentDir, { withFileTypes: true });
      for (const entry of entries) {
        if (entry.isDirectory()) {
          if (allIgnore.has(entry.name) || entry.name.startsWith('.')) continue;
          walk(path.join(currentDir, entry.name));
        } else if (entry.name.endsWith(extension)) {
          results.push(path.join(currentDir, entry.name));
        }
      }
    } catch { /* ignore permission errors */ }
  }

  walk(dir);
  return results;
}

// ---------------------------------------------------------------------------
// relativePath
// ---------------------------------------------------------------------------

/**
 * Get a relative path from a base directory, normalised with forward slashes.
 *
 * @param {string} base - absolute base directory
 * @param {string} filePath - absolute file path
 * @returns {string} - relative path with forward slashes
 */
function relativePath(base, filePath) {
  return path.relative(base, filePath).replace(/\\/g, '/');
}

// ---------------------------------------------------------------------------
// readFileSafe
// ---------------------------------------------------------------------------

/**
 * Read a file as UTF-8 string. Returns null on any error.
 *
 * @param {string} filePath - absolute path to file
 * @returns {string|null}
 */
function readFileSafe(filePath) {
  try {
    return fs.readFileSync(filePath, 'utf-8');
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// inferCommonFolder
// ---------------------------------------------------------------------------

/**
 * Detect the most common parent folder from a list of relative file paths.
 * Useful for pattern inference (e.g., "most entities live in Domain/Entities/").
 *
 * @param {string[]} filePaths - relative paths (forward slashes)
 * @returns {string|null} - most common parent folder with trailing slash, or null
 */
function inferCommonFolder(filePaths) {
  if (!filePaths.length) return null;

  const counts = new Map();
  for (const fp of filePaths) {
    const dir = path.dirname(fp).replace(/\\/g, '/');
    counts.set(dir, (counts.get(dir) || 0) + 1);
  }

  let maxDir = null;
  let maxCount = 0;
  for (const [dir, count] of counts) {
    if (count > maxCount) {
      maxDir = dir;
      maxCount = count;
    }
  }

  return maxDir ? maxDir + '/' : null;
}

module.exports = { collectFiles, relativePath, readFileSafe, inferCommonFolder, parseGitignoreDirs, DEFAULT_IGNORE };

File: apps/rt/src/run/scan/file_utils.rs
//! File-collection and path helpers shared across scanners — a port of
//! `registry/file-utils.js`.
//!
//! Only filesystem utilities live here: no scanning logic, no schema building.
//! Every function is fail-open (an unreadable directory yields an empty result,
//! never an error), matching the JS module's `try { … } catch { … }` shape.

use mustard_core::fs as mfs;
use std::collections::BTreeSet;
use std::path::{Path, PathBuf};

/// Universal directory skip-list — mirrors `DEFAULT_IGNORE` in `file-utils.js`.
pub const DEFAULT_IGNORE: &[&str] = &[
    "node_modules",
    "bin",
    "obj",
    "dist",
    ".next",
    "__pycache__",
    ".venv",
    "venv",
    "target",
    "build",
    ".git",
    "migrations",
    "Migrations",
];

/// Extract directory-name patterns from a `.gitignore` string.
///
/// A faithful port of `parseGitignoreDirs()` — conservative: keeps only entries
/// that look like a plain folder name (non-empty, no whitespace, no glob chars,
/// no slashes, not a negation, not a comment). Trailing slashes are stripped.
#[must_use]
pub fn parse_gitignore_dirs(content: &str) -> Vec<String> {
    let mut out = Vec::new();
    for raw_line in content.lines() {
        let line = raw_line.trim();
        if line.is_empty() || line.starts_with('#') || line.starts_with('!') {
            continue;
        }
        if line.starts_with('/') {
            continue; // path-anchored
        }
        if line
            .chars()
            .any(|c| c.is_whitespace() || matches!(c, '*' | '?' | '[' | ']'))
        {
            continue; // glob or whitespace
        }
        let name = line.strip_suffix('/').unwrap_or(line);
        if name.contains('/') {
            continue; // nested path, not a bare name
        }
        out.push(name.to_string());
    }
    out
}

/// Build the merged skip-set for a walk rooted at `dir`.
///
/// Combines `DEFAULT_IGNORE`, the explicit `ignore` argument, the
/// `MUSTARD_SCAN_IGNORE` env var (comma-separated), and directory entries
/// parsed from the subproject's `.gitignore` — exactly the four sources
/// `collectFiles` merges.
fn ignore_set(dir: &Path, ignore: &[&str]) -> BTreeSet<String> {
    let mut set: BTreeSet<String> = DEFAULT_IGNORE.iter().map(|s| (*s).to_string()).collect();
    for extra in ignore {
        set.insert((*extra).to_string());
    }
    if let Ok(env) = std::env::var("MUSTARD_SCAN_IGNORE") {
        for name in env.split(',') {
            let trimmed = name.trim();
            if !trimmed.is_empty() {
                set.insert(trimmed.to_string());
            }
        }
    }
    if let Ok(content) = mfs::read_to_string(&dir.join(".gitignore")) {
        for name in parse_gitignore_dirs(&content) {
            set.insert(name);
        }
    }
    set
}

/// Recursively collect every file with `extension` under `dir`.
///
/// Skips ignored directories, dot-directories, and the `ignore` argument —
/// a faithful port of `collectFiles()`. `extension` includes the dot
/// (e.g. `.rs`). Fail-open: unreadable directories are silently skipped.
#[must_use]
pub fn collect_files(dir: &Path, extension: &str, ignore: &[&str]) -> Vec<PathBuf> {
    let skip = ignore_set(dir, ignore);
    let mut results = Vec::new();
    walk(dir, extension, &skip, &mut results);
    results
}

fn walk(current: &Path, extension: &str, skip: &BTreeSet<String>, results: &mut Vec<PathBuf>) {
    let Ok(entries) = mfs::read_dir(current) else {
        return;
    };
    for entry in entries {
        let name: &str = &entry.file_name;
        if entry.is_dir {
            if skip.contains(name) || name.starts_with('.') {
                continue;
            }
            walk(&entry.path, extension, skip, results);
        } else if name.ends_with(extension) {
            results.push(entry.path);
        }
    }
}

/// Relative path from `base` to `file_path`, normalised with forward slashes.
///
/// A faithful port of `relativePath()`.
#[must_use]
pub fn relative_path(base: &Path, file_path: &Path) -> String {
    let rel = file_path.strip_prefix(base).unwrap_or(file_path);
    rel.to_string_lossy().replace('\\', "/")
}

/// Read a file as UTF-8, returning `None` on any error — a port of `readFileSafe()`.
#[must_use]
pub fn read_file_safe(file_path: &Path) -> Option<String> {
    mfs::read_to_string(file_path).ok()
}

/// Most common parent folder across a list of relative file paths.
///
/// A faithful port of `inferCommonFolder()` — returns the most frequent parent
/// directory with a trailing slash, or `None` for an empty input.
#[must_use]
pub fn infer_common_folder(file_paths: &[String]) -> Option<String> {
    if file_paths.is_empty() {
        return None;
    }
    let mut counts: Vec<(String, usize)> = Vec::new();
    for fp in file_paths {
        let normalized = fp.replace('\\', "/");
        let dir = match normalized.rfind('/') {
            Some(idx) => normalized[..idx].to_string(),
            None => ".".to_string(),
        };
        if let Some(entry) = counts.iter_mut().find(|(d, _)| *d == dir) {
            entry.1 += 1;
        } else {
            counts.push((dir, 1));
        }
    }
    counts
        .into_iter()
        .max_by_key(|(_, count)| *count)
        .map(|(dir, _)| format!("{dir}/"))
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn parse_gitignore_keeps_plain_names_only() {
        let content = "# comment\nnode_modules\n/anchored\n*.log\n  \nvendor/\n!keep\nsrc/nested\n";
        assert_eq!(
            parse_gitignore_dirs(content),
            vec!["node_modules".to_string(), "vendor".to_string()]
        );
    }

    #[test]
    fn collect_files_skips_ignored_dirs() {
        let dir = tempdir().unwrap();
        std::fs::write(dir.path().join("a.rs"), "").unwrap();
        std::fs::create_dir(dir.path().join("target")).unwrap();
        std::fs::write(dir.path().join("target").join("b.rs"), "").unwrap();
        std::fs::create_dir(dir.path().join("src")).unwrap();
        std::fs::write(dir.path().join("src").join("c.rs"), "").unwrap();

        let mut found: Vec<String> = collect_files(dir.path(), ".rs", &[])
            .iter()
            .map(|p| p.file_name().unwrap().to_string_lossy().to_string())
            .collect();
        found.sort();
        assert_eq!(found, vec!["a.rs".to_string(), "c.rs".to_string()]);
    }

    #[test]
    fn infer_common_folder_picks_most_frequent() {
        let paths = vec![
            "src/domain/user.rs".to_string(),
            "src/domain/order.rs".to_string(),
            "src/api/route.rs".to_string(),
        ];
        assert_eq!(infer_common_folder(&paths), Some("src/domain/".to_string()));
    }

    #[test]
    fn infer_common_folder_empty_is_none() {
        assert_eq!(infer_common_folder(&[]), None);
    }
}
